sktime/transformations/series/temporian.py

```diff
@@ -0,0 +1,156 @@
+# copyright: sktime developers, BSD-3-Clause License (see LICENSE file)
+
+"""Implementation of TemporianTransformer. Contributed by the Temporian team."""
+
+__author__ = ["ianspektor", "javiber"]
+
+from sktime.transformations.base import BaseTransformer
+
+
+class TemporianTransformer(BaseTransformer):
+    """Applies a Temporian function to the input time series.
+
+    This transformer applies a Temporian [1]_ function to the input time series.
+
+    The conversion from sktime's internal representation to Temporian's EventSet [2]_
+    and back is handled automatically by the transformer.
+
+    Parameters
+    ----------
+    function : Callable[[temporian.EventSet], temporian.EventSet]
+        Temporian function to apply to the input time series. The function must receive
+        and return a single Temporian EventSet, and can apply an arbitrary number of
+        Temporian operators to its input.
+    compile : bool, default=False
+        If True, the function will be compiled using Temporian's @tp.compile [3]_
+        decorator, which can lead to significant speedups by optimizing the graph of
+        operations.
+
+    Examples
+    --------
+    >>> from sktime.datasets import load_airline
+    >>> from sktime.transformations.series.temporian import TemporianTransformer
+    >>> import temporian as tp  # doctest: +SKIP
+    >>>
+    >>> def function(evset):  # doctest: +SKIP
+    >>>    return evset.simple_moving_average(tp.duration.days(3 * 365))  \
+        # doctest: +SKIP
+    >>> transformer = TemporianTransformer(function=function)  # doctest: +SKIP
+    >>> X = load_airline()  # doctest: +SKIP
+    >>> X_averaged = transformer.fit_transform(X)  # doctest: +SKIP
+
+    References
+    ----------
+    .. [1] https://temporian.readthedocs.io/en/stable/
+    .. [2] https://temporian.readthedocs.io/en/stable/reference/temporian/EventSet/
+    .. [3] https://temporian.readthedocs.io/en/stable/reference/temporian/compile/
+    """
+
+    _tags = {
+        # packaging info
+        # --------------
+        "authors": ["ianspektor", "javiber"],
+        "maintainers": ["ianspektor", "javiber"],
+        "python_dependencies": ["temporian"],
+        "python_version": ">=3.8",
+        # estimator tags
+        # --------------
+        "univariate-only": False,
+        "scitype:transform-input": "Series",
+        "scitype:transform-output": "Series",
+        "scitype:instancewise": True,
+        "scitype:transform-labels": "None",
+        "X_inner_mtype": "pd.DataFrame",
+        "fit_is_empty": True,
+        "capability:inverse_transform": False,
+        "capability:unequal_length": True,
+        "handles-missing-data": False,
+    }
+
+    def __init__(self, function, compile=False):
+        import temporian as tp
+
+        self.function = function
+        self.compile = compile
+        if compile:
+            self.function = tp.compile(self.function)
```

Comment:
yes, I think that's an interface violation. The function should be cloned or copied.

The tests do not detect that since for arbitrary objects, checking "identity" before vs after becomes problematic if an unknown type is encountered. For objects, we check identity of reference, but we do not check anything further unless it is an skbase-compatible estimator. This is typically not a problem, the only case are adaptors to non-skbase dependent libraries, which are very rare - but exist, like this PR.

Not sure how to cover this in the tests, but it is an excellent spot!

Reply:
note that `tp.compile` doesn't modify `self.function` (it runs the graph inside it and returns a different function). the original function is left intact. what exactly is the problem/what changes should we make?

Reply:
Something like this?

```python
self._function = tp.compile(self.function) if compile else self.function
```

And rest of the code should only use self._function only.

---

I believe this is the contract @fkiraly always refer to, and unfortunately I'm lazy enough to ever read it fully.

````

...

In addition, every keyword argument accepted by __init__ should correspond to an attribute on the instance. Scikit-learn relies on this to find the relevant attributes to set on an estimator when doing model selection.

To summarize, an __init__ should look like:

```
def __init__(self, param1=1, param2=2):
    self.param1 = param1
    self.param2 = param2
```

There should be no logic, not even input validation, and the parameters should not be changed. The corresponding logic should be put where the parameters are used, typically in fit.

...
````

Ref. https://scikit-learn.org/stable/developers/develop.html#instantiation

Reply:
Oh. According to that then we should be compiling the function during fit, not on init?

Reply:
Moved the compliation to `fit()`, `__init__` now follows the guidelines above 👍🏼